src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,25 +1,76 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
+using System.Diagnostics;
+
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class VersionConverter : JsonConverter<Version>
     {
         public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            if (reader._stringHasEscaping)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+#if BUILDING_INBOX_LIBRARY
+
+            ReadOnlySpan<byte> source = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
```

Comment:
If you're accessing the raw bytes and performing the conversion to chars yourself, this could introduce a security vulnerability because it could allow invalid JSON through the system. `Version.TryParse` accepts a wider range of characters than the JSON spec allows.

@layomia as FYI, since you'll recognize this as the same feedback I gave for the recent `TimeSpan` optimizations. This is now the second PR I've seen that uses this pattern. We need to go through all the optimizations that were checked in as part of the 6.0 wave and ensure this was not actually committed anywhere.

Reply:
> Version.TryParse accepts a wider range of characters than the JSON spec allows

Example? Is this about whitespace?

Reply:
Not just whitespace, but nulls and other weirdness.

```cs
string s = "\n10\0.\v20\t\0\0.\r\n30\0"; // lots of invalid JSON
Version v = new Version(s);
Console.WriteLine(v);
```

Reply:
> Not just whitespace, but nulls and other weirdness

This is what I meant by whitespace.  Int32.{Try}Parse by default allows for a subset of what char.IsWhitespace considers valid, and also ignores trailing nulls.  Version.TryParse just inherits that.

What, specifically, is the security concern with not having validation fail because of such characters?

Reply:
We make a guarantee that using default deserialization parameters and built-in converters, the incoming JSON payload is fully validated before a full object graph can be hydrated. This guarantee exists to provide support for heterogeneous environments. For instance: you have a frontend performing an initial pass over the input data, including basic consistency checks and request authentication; then that payload is forwarded to a backend for real processing. The frontend should not allow non-compliant data to reach the backend, as the backend might not be resilient against processing that data, or it might store the data in a manner that corrupts the database. (There were vulnerabilities in DasBlog, aspnet 4.x, and Exchange due to two parsers in a heterogeneous environment interpreting the same payload in different manners.)

One quasi-by-design hole in the _System.Text.Json_ stack is that string data is not validated until the string itself is materialized. For most parsers and converters this behavior is just fine, as they get the string (which validates!), then parse it to create the final result. But in the case of hyper-optimized converters like the one under discussion here, they're dropping down to the raw byte level to get at the underlying data and avoid string materialization, which can suppress the normal correctness guarantees we make for the user.

Reply:
In proposed version i added [failing fast if string has any escaping](https://github.com/dotnet/runtime/blob/07b6a9af6048817981251f7f67bb085870db97f7/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs#L13-L16).  
If string data with escaping is still valid Version - then this failing should be removed, and replaced with unescaping in case if string has any escaping. This way byte data will basically follow the same path as when you call `reader.GetString` except calling `Encoding.GetChars` instead of `Encoding.GetString` in the end. But I can't reproduce validation on materialization that string does, that you are talking about, with the example invalid json you presented earlier.

```cs
string source = "\n10\0.\v20\t\0\0.\r\n30\0"; // lots of invalid JSON
var s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
byte[] ut8Bytes = s_utf8Encoding.GetBytes(source);
string convertedString = s_utf8Encoding.GetString(ut8Bytes);
Span<char> chars = stackalloc char[source.Length];
s_utf8Encoding.GetChars(ut8Bytes, chars);
Console.WriteLine(chars.SequenceEqual(convertedString) && chars.SequenceEqual(source)); // Prints true
Version versionFromSource = new Version(source);
Version versionFromConvertedStr = new Version(convertedString);
Version versionFromChars = Version.Parse(chars);
Console.WriteLine(versionFromSource.Equals(versionFromConvertedStr) && versionFromSource.Equals(versionFromChars) 
                    && versionFromConvertedStr.Equals(versionFromChars)); // Prints true
```

Reply:
@GrabYourPitchforks, thanks, I get all that, but I'm not clear how that's applicable to this specific case.  We're not talking about taking unvalidated json and storing it somewhere that expects validated json; we're talking about ending up with a well-formed Version object that discarded the unnecessary whitespace/nulls.  I'm not saying that's not divergent from the json spec, but I'm also not seeing how it results in a vulnerability.

Reply:
For consistency, we're taking the approach of disallowing all leading or trailing trivia across all the primitive types that we support across the STJ stack. This involves implementing all the necessary validation prior to calling an underlying lower-level API that does the actual data parsing, e.g. `Utf8Parser`. This has been a manual process repeated for each supported data type, and we'll do the same for this PR.

> please schedule some time to audit all the optimizations that were checked in as part of the 6.0 wave

@GrabYourPitchforks the `TimeSpan` PRs (https://github.com/dotnet/runtime/pull/54186, https://github.com/dotnet/runtime/pull/55350) and this PR are the only candidates for .NET 6 where we process data types at this level so I think these code reviews might suffice. 

Here are some somewhat related PRs in case something jumps out:
- https://github.com/dotnet/runtime/pull/54254
- https://github.com/dotnet/runtime/pull/51367